container_api/src/container.rs

```diff
@@ -42,10 +72,11 @@ use interface::{ContainerApiBuilder, InstanceMap, Interface};
 /// and also enable easier testing, a DnaLoader ()which is a closure that returns a
 /// Dna object for a given path string) has to be injected on creation.
 pub struct Container {
-    instances: InstanceMap,
-    config: Configuration,
-    interface_threads: HashMap<String, InterfaceThreadHandle>,
-    dna_loader: DnaLoader,
+    pub(crate) instances: InstanceMap,
+    pub(crate) config: Configuration,
+    pub config_path: PathBuf,
+    interface_threads: HashMap<String, Sender<()>>,
+    pub(crate) dna_loader: DnaLoader,
```

Comment:
Well, they are *just* crate public.. I guess we also make them only `ContainerAdmin` public.. but that, I guess, would be easier to reason about than having setters.


I regard `ContainerAdmin` as part of the container but wanted to create another bucket instead of putting all new functions into the already quite big `struct Container` and file container.rs.

Reply:
`Yeah `pub(crate)` helps, but it still means we have to consider every file in the crate (better than every file in the code base at least) to understand how those members are being used. IMO it's always useful to spend the small bit of energy upfront to avoid this, if it is indeed a small amount of energy. It's so much harder to go back to private later once we've leaked a member out into the public. IMO ideally every public member should have a *very* good reason for being publicly mutable.

Reply:
I tried `pub(in crate::container_admin::ContainerAdmin)` but that didn't work, I think because the other module needs to be parent.. So the alternative would be merging `ContainerAdmin` into container which I think is making it harder to read than this pub(crate).. I'm very open to suggestions :)

Reply:
I haven't looked into why these things need to be referenced outside of the container but would it be possible to just write getters and setters?

Reply:
Setters would make those things writable to everybody which IMHO is worse than making them crate public.

Needed because I've put the container admin functions in a trait `ContainerAdmin` that gets implemented for `Container`.